### PR TITLE
Feedback on regex parser PR

### PR DESCRIPTION
**Pull request by danmoseley**
Feedback from https://github.com/dotnet/runtime/pull/88558

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 1]
```diff
@@ -521,19 +521,17 @@ private RegexNode ScanReplacement()
                 int startpos = _pos;
 
                 _pos = _pattern.IndexOf('$', _pos);
-                if (_pos == -1)
+                if (_pos < 0)
                     _pos = _pattern.Length;
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:525]:
```C#
if (_pos < 0)
{
    _pos = _pattern.Length;
}
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 2]
```diff
@@ -521,19 +521,17 @@ private RegexNode ScanReplacement()
...
                 {
-                    if (_pattern[_pos++] == '$')
-                    {
-                        _unit = ScanDollar();
-                    }
+                    _pos++;
+                    _unit = ScanDollar();
 
-                    _concatenation.AddChild(_unit!);
+                    _concatenation.AddChild(_unit);
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:534]: Can this be:
```C#
_pos++;
_concatenation.AddChild(ScanDollar());
_unit = null;
```
?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 3]
```diff
@@ -1052,13 +1050,13 @@ private void ScanBlank()
                 if ((_options & RegexOptions.IgnorePatternWhitespace) != 0 && _pos < _pattern.Length && _pattern[_pos] == '#')
                 {
                     _pos = _pattern.IndexOf('\n', _pos);
-                    if (_pos == -1)
+                    if (_pos < 0)
                         _pos = _pattern.Length;
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1054]: braces

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs [hunk 4]
```diff
@@ -1480,14 +1482,12 @@ private char ScanHex(int c)
...
                 {
-                    int d;
                     char ch = _pattern[_pos++];
-                    if ((uint)(d = ch - '0') <= 9)
-                        i = (i * 0x10) + d;
-                    else if ((uint)(d = (ch | 0x20) - 'a') <= 5)
-                        i = (i * 0x10) + d + 0xa;
-                    else
+                    int result = HexConverter.FromChar(ch);
+                    if (result == 0xFF)
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1487]: braces

**Review by stephentoub — APPROVED** [review]
Thanks

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 3 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:1054]: ah, I wasn't aware of the coding guidelines that braces are required in a nested context

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs:534]: yes, changed, but in general the to and fro with _unit and _concatenation and _group etc can probably be cleaned up in future. Eg., I think each time you add a child to a concatenation, you null _unit, but all the callers do it. etc

**Comment by danmoseley** [comment]
``` 

cases)
  Starting:    System.Net.NameResolution.Functional.Tests (parallel test collections = on, max threads = 2)
Process terminated. Assertion failed.
   at System.Net.NetEventSource.Info(String thisOrContextObject, String memberName, String message) in /_/src/libraries/Common/src/System/Net/Logging/NetEventSource.Common.cs:line 80
   at System.Net.NetEventSource.Info(Object thisOrContextObject, FormattableString formattableString, String memberName) in /_/src/libraries/Common/src/System/Net/Logging/NetEventSource.Common.cs:line 67
   at System.Net.Dns.<>c.<GetHostEntryAsync>b__6_0(Task`1 t, Object s) in /_/src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:line 127
   at System.Threading.Tasks.ContinuationTaskFromResultTask`1.InnerInvoke() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TaskContinuation.cs:line 140
```

**Pull request merged**